### ISSUE TITLE
Advice on bulk reference parsing?

### ISSUE THREAD

**Issue by bnewbold** [kermitt2/grobid#792]
We have a use case at Internet Archive where we need to parse 100+ million bibliographic reference strings with GROBID into structured form. These are "free-floating" strings, coming from sources like Crossref, which have not been normalized (eg, mapped to a specific DOI), and have not been parsed or structured. We do not want to "consolidate" yet (eg, using biblio-glutton), just want to parse out title, year, volume, issue, etc into XML (which we will transform to JSON).

The question is whether there is any advice on tuning, or if patches to GROBID would be helpful for this specific use-case. The "Process Citation" REST API endpoint should work, but it feels like there is a lot of overhead making a separate REST request for each individual citation (even in parallel, say 30 requests in flight concurrently), instead of submitting all the citations for a paper at a time (eg, a batch of 50, maybe separated by newlines or as separate fields). I don't know enough about GROBID internals to understand if these would need to be run through the citation parsing model one-at-a-time anyways.

I will try to report back on this issue with an estimate of throughput (parsed citations/second using a large-ish server).

**Comment by kermitt2** [comment]
Hi @bnewbold !

I have the same impression, using only the "Process Citation" REST API endpoint for millions of single reference strings might introduce a lot of overhead just managing the concurrent requests. 

The other aspect is that for reference string parsing it's better to use the deep learning model, which performs better (it's the only Deep Learning model really performing better than CRF). When using a deep learning model, we want to process reference strings in batch too, to take advantage of the GPU parallelization. GPU parallelization is well managed when processing a PDF (all the reference strings of a PDF are batched for the deep learning model), but the individual "Process Citation" REST call, they will remain single per batch, which will make runtime performance abyssal in this use case. 

What would you prefer:

- a batch command line `processBatchReferences` which takes a txt file with one reference string per line and manage the parallel processing (including deep learning batch)
- a REST API end point `processCitations` which takes a list of reference strings (list of arbitrary lenght) and manage the deep learning batch

In both cases, it's just cut and place of code doing this at other places.

**Comment by bnewbold** [comment]
For us, an API endpoint is best, that lets us do simple HTTP(S) load balancing against a cluster of workers. Maybe `processCitationList` or `processCitationSet` so it isn't a single-character difference from the existing endpoint?

Do you think that the type article size of references (~50 refs) is reasonably close to the optimal batch size for GPU parallelization? If not, and larger (or smaller) batch sizes are better, we can still re-batch sets in our client code and submit to the API that way (eg, groups of 1000 reference strings or something like that).

My expectation is that the API endpoint would return TEI-XML with structure similar to that of a full article parse (process fulltext), and the references would have arbitrary keys/indexes invented ('b1', 'b2', 'b3' or whatever is the default). "consolidate or not" and "include raw/unpased reference" as with the existing endpoint would be helpful. Not sure if the request should be a single string field with newline-delimited references or multiple fields or what. IIRC all the other API endpoints are form encoded and I can't remember if there is a standard pattern for an array of strings.

**Comment by kermitt2** [comment]
Ok for `processCitationList` then (because the order in the response will be maintained) !

So to clarify, with a web API endpoint, it will be the task of the client to manage the concurrent calls, as for the other grobid service. So if your large-ish server has 64 threads, you would need to set in the config file `concurrency` to 64 for instance, and then have your client sending 64 requests at least (a bit more) in parallel (watching some response `503` to wait a bit for more for re-sending the request). 

For the size of the reference list per request, I would say intuitively it's going to perform better with large sets, like 1000 or more.  GPU batch size for citation BidLSTM-CRF-FEATURES model is 20, so anything larger will be good, and multiple of 20 to optimize. 

I am not super sure how GPU will work for very high parallelization. For the citation model, I've never tested with more than 8 parallel threads with one GPU, which works almost as fast as CRF with 8 threads when processing complete documents - when  processing only references it will be much slower I think. Maybe with one GPU we will have a bottleneck after a certain level of concurrency. I've never been able to test DeLFT with multiple GPU and I assume it is not supported. One solution could be to start several GROBID docker image with a different assigned GPU and port, and then distribute the queries to the different services. 

With CRF, I think we should be around 180-200 references processed per second with 1 thread, so maybe possible to process 100M references in around one day with 8-10 threads. 

> My expectation is that the API endpoint would return TEI-XML with structure similar to that of a full article parse (process fulltext), and the references would have arbitrary keys/indexes invented ('b1', 'b2', 'b3' or whatever is the default). "consolidate or not" and "include raw/unpased reference" as with the existing endpoint would be helpful. 

Yes, we keep the same parameters as the other services for consistency.

> Not sure if the request should be a single string field with newline-delimited references or multiple fields or what. IIRC all the other API endpoints are form encoded and I can't remember if there is a standard pattern for an array of strings.

Yes I think this is standard, the usual javascript, java, etc. web client supports that transparently. The problem I had with this is with curl query, we need to repeat the form parameters arguments for every elements of the list... but nobody will use a curl request for this :D ?

**Comment by kermitt2** [comment]
@bnewbold This is available in branch https://github.com/kermitt2/grobid/pull/798
It can be tested with the python client, branch https://github.com/kermitt2/grobid_client_python/tree/processCitationList

**Comment by bnewbold** [comment]
I have started playing with this, but haven't finished a proper benchmark quite yet.

In Python, trying `aiohttp` instead of `requests` or `grobid_client_python`, I had to pass the `citations` form field multiple times, which I was able to do by passing a list-of-tuples (key, value) instead of a dict. Maybe `requests` handled that automatically? I think it is probably worth mentioning in the API doc.

I noticed that the default with no `Accept` header was to return BibTeX, which I think is a change in behavior? But maybe that applies to all endpoints now since the older version of GROBID I have been running. Will need to experiment more.

Noted that this is only a `POST` endpoint, with no `PUT` option, because of the need for repeated form field.

**Comment by kermitt2** [comment]
> In Python, trying `aiohttp` instead of `requests` or `grobid_client_python`, I had to pass the `citations` form field multiple times, which I was able to do by passing a list-of-tuples (key, value) instead of a dict. Maybe `requests` handled that automatically? I think it is probably worth mentioning in the API doc.

Weird, according to the `aiohttp` documentation, for the form-encoded data, "a dictionary to the data argument" must be passed:
https://docs.aiohttp.org/en/stable/client_quickstart.html#more-complicated-post-requests

It is strange to repeat the parameters for every list value of the `citations` field. I hope `aiohttp` is not sending then multiple independent async `processCitationList` request for each reference behind the scene !

I guess there are so many different languages and web client frameworks, it really depends on each regarding list value for form-encoded POST queries and beyond a reference examples like in `grobid_client_python`, we probably don't want to document them individually.

> I noticed that the default with no Accept header was to return BibTeX, which I think is a change in behavior? But maybe that applies to all endpoints now since the older version of GROBID I have been running. Will need to experiment more.

No it's not the case, `Accept: application/x-bibtex` must be explicitely set to return BibTeX stuff:

```
$ curl -X POST -H "Accept: " -d "citations=Graff, Expert. Opin. Ther. Targets (2002) 6(1): 103-113" -d "citations=Groff, Expert. Opin. Ther. Targets (2002) 6(1): 103-113" localhost:8070/api/processCitationList

<TEI xmlns="http://www.tei-c.org/ns/1.0" xmlns:xlink="http://www.w3.org/1999/xlink" 
 xmlns:mml="http://www.w3.org/1998/Math/MathML">
	<teiHeader/>
	<text>
		<front/>
		<body/>
		<back>
			<div>
				<listBibl>
<biblStruct xml:id="b0">
	<analytic>
		<title/>
		<author>
			<persName><surname>Graff</surname></persName>
		</author>
	</analytic>
	<monogr>
		<title level="j">Expert. Opin. Ther. Targets</title>
		<imprint>
			<biblScope unit="volume">6</biblScope>
			<biblScope unit="issue">1</biblScope>
			<biblScope unit="page" from="103" to="113" />
			<date type="published" when="2002" />
		</imprint>
	</monogr>
</biblStruct>

<biblStruct xml:id="b1">
	<analytic>
		<title/>
		<author>
			<persName><surname>Groff</surname></persName>
		</author>
	</analytic>
	<monogr>
		<title level="j">Expert. Opin. Ther. Targets</title>
		<imprint>
			<biblScope unit="volume">6</biblScope>
			<biblScope unit="issue">1</biblScope>
			<biblScope unit="page" from="103" to="113" />
			<date type="published" when="2002" />
		</imprint>
	</monogr>
</biblStruct>

				</listBibl>
			</div>
		</back>
	</text>
</TEI>
```

```
$ curl -X POST -H "Accept: application/x-bibtex" -d "citations=Graff, Expert. Opin. Ther. Targets (2002) 6(1): 103-113" -d "citations=Griff, Expert. Opin. Ther. Targets (2002) 6(1): 103-113" localhost:8070/api/processCitationList

@article{0,
  author = {Graff},
  journal = {Expert. Opin. Ther. Targets},
  year = {2002},
  pages = {103--113},
  volume = {6},
  number = {1}
}

@article{1,
  author = {null, Griff},
  journal = {Expert. Opin. Ther. Targets},
  year = {2002},
  pages = {103--113},
  volume = {6},
  number = {1}
}
```

> Noted that this is only a POST endpoint, with no PUT option, because of the need for repeated form field.

No I was just lazy :) 
I could add the `PUT` option I suppose, but I don't know if it is of any use given that we already have a `POST` ?
I was actually thinking to remove all the existing `PUT` support (the GROBID web API is super old, I don't remember why we add a `PUT`), because I don't see the use and it is not consistent with the `PUT` meaning.

**Comment by bnewbold** [comment]
Took me a long time to get back to this. From informal benchmarking on my laptop (with 4 vCPUs, running Debian/Linux, no containerization, no GPU, CRF not DeLFT), I found that:

- multiple client requests in parallel, with processCitations (not List), is a roughly linear speedup, as would be expected
- processCitationList with batch size of 1 was same performance as processCitations, as would be expected
- throughput, as in "total wall time to process 8,000 citations", was about 20% faster (80% of wall time) with batch sizes of 20 or greater, using client parallel requests = number of CPUs. larger batch sizes, up to 500, did not seem to make a measurable performance improvement. batch sizes of 10 or fewer were almost as fast.

Unfortunately I haven't had time to test with a GPU/DeFLT (we will probably be sticking with CPUs for now). The batch size performance may scale differently when using those models, which could make an even bigger performance impact for other folks.

This is pretty great! And being able to submit a batch of citations at once makes some client code much simpler.

--- 

Regarding POST form requests, I don't think documenting behavior for specific client libraries is necessary, but I think noting how the request should be structured at the HTTP level would be good. Eg, that the request should be `application/x-www-form-urlencoded` or `multipart/form-data`, and that the 'citations' key should be repeated for each separate citation value. As opposed to, eg, a single `citations` with value being a one-citation-per-line text file. I remember reading the `processCitation` docs in the past, and the 'citations' key (plural) made me think that I should pass a single string with newline-delimited citations.

To propose something specific, text could be:

> bibliographical references to be parsed, as raw strings, with form field repeated for each separate citation

to replace:

> bibliographical reference to be parsed as a list of raw strings

---

I couldn't recreate the "bibtext-as-default" thing, maybe I copy-pasted something by accident. Sorry for the noise.

**Comment by bnewbold** [comment]
I tried to do some quick testing on an AWS EC2 instance with a GPU (`g4dn.4xlarge`, with Nvidia Tesla T4), but didn't seem to be able to get DeLFT to actually use the GPU.

I followed the GROBID docker docs, reconfigued all the commented out 'delft' backends, and after rebuilding a SNAPSHOT image, during startup GROBID said:

```
    INFO  [2021-08-19 06:29:43,172] org.grobid.core.jni.DeLFTModel: Loading DeLFT model for citation with architecture BidLSTM_CRF_FEATURES...
    running thread: 1
    INFO  [2021-08-19 06:29:43,174] org.grobid.core.jni.JEPThreadPool: Creating JEP instance for thread 23
    Using TensorFlow backend.
    WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/delft/sequenceLabelling/preprocess.py:24: The name tf.set_random_seed is deprecated. Please use tf.compat.v1.set_random_seed instead.
    WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/delft/sequenceLabelling/wrapper.py:22: The name tf.logging.set_verbosity is deprecated. Please use tf.compat.v1.logging.set_verbosity instead.
    WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/delft/sequenceLabelling/wrapper.py:22: The name tf.logging.ERROR is deprecated. Please use tf.compat.v1.logging.ERROR instead.
    loading model weights /opt/grobid/grobid-home/models/citation-BidLSTM_CRF_FEATURES/model_weights.hdf5
```

which seemed promising. Then ran `nvidia-smi` and got:

```
    Thu Aug 19 07:02:48 2021
    +-----------------------------------------------------------------------------+
    | NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
    |-------------------------------+----------------------+----------------------+
    | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
    | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
    |                               |                      |               MIG M. |
    |===============================+======================+======================|
    |   0  Tesla T4            On   | 00000000:00:1E.0 Off |                    0 |
    | N/A   43C    P0    27W /  70W |    102MiB / 15109MiB |      0%      Default |
    |                               |                      |                  N/A |
    +-------------------------------+----------------------+----------------------+

    +-----------------------------------------------------------------------------+
    | Processes:                                                                  |
    |  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
    |        ID   ID                                                   Usage      |
    |=============================================================================|
    |    0   N/A  N/A     42196      C   java                               99MiB |
    +-----------------------------------------------------------------------------+
```

which was also promising, that the java process seemed connected. However, actual requests were pretty slow and used a lot of CPU, and the reported GPU utilization never went up.

Elsewhere I read that tensorflow 1.15.x does not support CUDA 11? Maybe that is the issue? Anyways, this is probably a separate issue, and I have already torn down this VM.

**Comment by kermitt2** [comment]
Thanks a lot for the feedback @bnewbold !

The documentation of the PR specifies only `multipart/form-data`, but I still don't really understand the following: 

> bibliographical references to be parsed, as raw strings, with form field repeated for each separate citation

Maybe I am wrong, but I think the expected and normal usage with a client is to pass the list of citations as a list? 

It works like this in python with `requests`: 

https://github.com/kermitt2/grobid_client_python/blob/processCitationList/grobid_client/grobid_client.py#L305 
(this is then mapped to `requests.request()`)

With aiohttp, I quickly test the following, which works fine: 

```
import aiohttp
import asyncio

async def main():
    async with aiohttp.ClientSession() as session:

        payload = {'citations': ['Newbold, Bryan, "My first reference", 2020.', 'Lopez, Patrice. "my second reference", 2021.'] }
        async with session.post('http://localhost:8070/api/processCitationList', data=payload) as resp:
            print(await resp.text())

loop = asyncio.get_event_loop()
loop.run_until_complete(main())
```

```

<TEI xmlns="http://www.tei-c.org/ns/1.0" xmlns:xlink="http://www.w3.org/1999/xlink" 
 xmlns:mml="http://www.w3.org/1998/Math/MathML">
	<teiHeader/>
	<text>
		<front/>
		<body/>
		<back>
			<div>
				<listBibl>
<biblStruct xml:id="b0">
	<monogr>
		<title level="m" type="main">My first reference</title>
		<author>
			<persName><forename type="first">Bryan</forename><surname>Newbold</surname></persName>
		</author>
		<imprint>
			<biblScope unit="page">2020</biblScope>
		</imprint>
	</monogr>
</biblStruct>

<biblStruct xml:id="b1">
	<monogr>
		<title level="m" type="main">my second reference</title>
		<author>
			<persName><forename type="first">Patrice</forename><surname>Lopez</surname></persName>
		</author>
		<imprint>
			<biblScope unit="page">2021</biblScope>
		</imprint>
	</monogr>
</biblStruct>

				</listBibl>
			</div>
		</back>
	</text>
</TEI>
```


In Java/Jersey, we can pass a list in POST via a parameter like `Map<String,List<String>> references` (it works for GET queries too with queryParam).

So normally existing web request API supports lists as parameter value well ?

> I remember reading the processCitation docs in the past, and the 'citations' key (plural) made me think that I should pass a single string with newline-delimited citations.

Yes I fully agree, it should be just `reference` for `processReference` (instead of `processCitation`) and `references` for  `processReferences`, but for legacy reason if I change this it will break a lot of things in many places for many users and I am not prepared to this :) , so I prefer to stick in this bad naming for the moment.

**Comment by kermitt2** [comment]
> Elsewhere I read that tensorflow 1.15.x does not support CUDA 11? Maybe that is the issue? Anyways, this is probably a separate issue, and I have already torn down this VM.

yes it looks like the reason. I tested the docker container with CUDA 9 and 10 and it works, but it seems that CUDA 11 gives this behaviour. So I will have to update DeLFT to TF 2.0 quickly, but we will loose the support of ELMo.

**Comment by kermitt2** [comment]
On the GPU side, are you using a docker image built yourself? One constraint of the nvidia container toolkit is that the docker image has to be built on a machine with CUDa and GPU capacities to have a GPU-capable container.

**Comment by lfoppiano** [comment]
@bnewbold check PR #833 should fix the issue
